OpenDreamShared/Resources/DMIParser.cs

```diff
@@ -44,6 +45,80 @@ public sealed class ParsedDMIDescription {
                 };
             }
 
+            public void InsertIcon(ParsedDMIDescription icon, string icon_state, AtomDirection? dir, int? frame, float? delay)
+            {
+                List<ParsedDMIState> states;
+                if (icon_state is null)
+                {
+                    states = icon.States.Values.ToList();
+                }
+                else
+                {
+                    states = new List<ParsedDMIState>(1) { icon.States[icon_state] };
+                }
+
+                foreach (var state in states)
+                {
+                    if (dir is not null)
+                    {
+                        var goodDir = state.Directions[dir.Value];
+                        state.Directions.Clear();
+                        state.Directions = new Dictionary<AtomDirection, ParsedDMIFrame[]>(1) { { dir.Value, goodDir } };
+                    }
+
+                    if (frame is not null)
+                    {
+                        // TODO Ref says it must start at 1, need to check behavior for when it's less. Manually validate it for now.
+                        var goodFrame = Math.Max(frame.Value, 1);
+                        foreach (var (direction, frames) in state.Directions)
+                        {
+                            state.Directions[direction] = new ParsedDMIFrame[1] { frames[goodFrame] };
+                            if (delay is not null)
+                            {
+                                if (delay > 0)
+                                {
+                                    state.Rewind = false;
+                                }
+                                else if(delay < 0)
+                                {
+                                    state.Rewind = true;
+                                }
+                                else
+                                {
+                                    if (States.Count == 0)
+                                    {
+                                        // TODO Delay should be 1 tick. Is this a tick?
+                                        frames[goodFrame].Delay = 1f;
+                                    }
+                                    else
+                                    {
+                                        // Delay of the nearest frame. Hopefully. Christ
+                                        var stateDirs = States.Last().Value.Directions;
+                                        frames[goodFrame].Delay = dir is null
+                                            ? stateDirs.First().Value.Last().Delay
+                                            : stateDirs[dir.Value].Last().Delay;
+                                    }
+                                }
+                            }
```

Comment:
This is an incorrect implementation of the `delay` arg. This arg does nothing when it is `0` or `null`, and otherwise sets the state to rewind if it is `< 0` and sets the frame's delay to its absolute value.
```suggestion
                            if (delay is not null or 0) {
                                state.Rewind = delay < 0;
                                frames[goodFrame].Delay = Math.Abs(delay.Value);
                            }
```